### PR TITLE
Add more mirrors for the D install script

### PR DESCRIPTION
**Pull request by wilzbach**
Motivation: Network connections can fail and sometimes do they. CI shouldn't fail though.

The installer script is pushed to three locations (https://github.com/dlang/installer/pull/287).

See also: https://github.com/travis-ci/travis-build/pull/910 (introduction of the first mirror)

CC @MartinNowak @klickverbot @ibuclaw

### CODE REVIEW

File: spec/build/script/d_spec.rb [hunk 1]
```diff
@@ -13,7 +13,7 @@
   it_behaves_like 'a build script sexp'
 
   it 'downloads and runs the installer script' do
-    should include_sexp [:cmd, %r{curl.*https://dlang\.org/install\.sh.*\|\|.*curl.*https://nightlies\.dlang\.org/install\.sh}m,
+    should include_sexp [:cmd, %r{curl.*https://dlang\.org/install\.sh.*\|\|.*curl.*https://nightlies\.dlang\.org/install\.sh.*\|\|.*curl.*https://nightlies\.dlang\.org/install\.sh.*\|\|.*curl.*https://raw.githubusercontent.com/dlang/installer.*}m,
```
Comment (BanzaiMan) [inline spec/build/script/d_spec.rb]: This looks to be wrong. https://travis-ci.org/travis-ci/travis-build/jobs/323798377#L673

**Comment by wilzbach** [comment]
An example of Travis failures due to the `install.sh` script missing:

https://travis-ci.org/d-widget-toolkit/dwt/jobs/325059281

(provided by @jacob-carlborg)

File: lib/travis/build/script/d.rb [hunk 2]
```diff
@@ -18,22 +19,34 @@ def setup
...
           sh.fold 'compiler-download' do
             sh.echo 'Installing compiler and dub', ansi: :yellow
 
             sh.cmd 'CURL_USER_AGENT="Travis-CI $(curl --version | head -n 1)"'
             sh.cmd <<-'EOT'.gsub(/^              /, '')
+              __mirrors=(
+                "https://dlang.org/install.sh"
+                "https://downloads.dlang.org/other/install.sh"
+                "https://nightlies.dlang.org/install.sh"
+                "https://raw.githubusercontent.com/dlang/installer/0ddce7a399eda9201db3eab90c6ad073d92ac1c2/script/install.sh"
```
Comment (jacob-carlborg) [inline lib/travis/build/script/d.rb]: Is this the best URL? Would `https://github.com/dlang/installer/raw/0ddce7a399eda9201db3eab90c6ad073d92ac1c2/script/install.sh` be more stable?

Comment (wilzbach) [reply lib/travis/build/script/d.rb]: I don't think that `raw.githubusercontent.com` is going to vanish soon.
The direct GitHub link would introduce another point of failure:

```
> curl -v https://github.com/dlang/installer/raw/0ddce7a399eda9201db3eab90c6ad073d92ac1c2/script/install.sh
<html><body>You are being <a href="https://raw.githubusercontent.com/dlang/installer/0ddce7a399eda9201db3eab90c6ad073d92ac1c2/script/install.sh">redirected</a>.</body></html>
```

Comment (jacob-carlborg) [reply lib/travis/build/script/d.rb]: Well, you would need to follow redirects using `-L`.

Comment (jacob-carlborg) [reply lib/travis/build/script/d.rb]: I'm pretty sure I've seen invalid URLs in the wild which point to raw GitHub content that have a different domain name.

**Comment by wilzbach** [comment]
More example of recent failures:

- https://travis-ci.org/dlang-bots/dlang-bot/jobs/326026283
- https://travis-ci.org/dlang/dmd/jobs/325906456
- https://travis-ci.org/dlang/dmd/jobs/325882158
- https://travis-ci.org/dlang/dmd/jobs/325716863
- https://travis-ci.org/dlang/dmd/jobs/325443953
- https://travis-ci.org/dlang/dmd/jobs/325288674
- https://travis-ci.org/dlang/dmd/jobs/325067512

File: lib/travis/build/script/d.rb (hunk 2 above)
Comment (wilzbach) [reply lib/travis/build/script/d.rb]: Fair enough. Done.

**Comment by jacob-carlborg** [comment]
LGTM 👌.

**Comment by wilzbach** [comment]
FYI: I also submitted a PR to the DLang installer script to make it more robust: https://github.com/dlang/installer/pull/291

File: lib/travis/build/script/d.rb [hunk 3]
```diff
@@ -18,22 +19,34 @@ def setup
...
+              )
               for i in {0..4}; do
-                  if curl -fsS -A "$CURL_USER_AGENT" --max-time 5 https://dlang.org/install.sh -O ||
-                          curl -fsS -A "$CURL_USER_AGENT" --max-time 5 https://nightlies.dlang.org/install.sh -O; then
-                      break
-                  elif [ $i -lt 4 ]; then
-                      sleep $((1 << $i))
-                  fi
+                for mirror in "${__mirrors[@]}" ; do
+                    if curl -fsSL -A "$CURL_USER_AGENT" --connect-timeout 5 --speed-time 30 --speed-limit 1024 "$mirror" -O ; then
```
Comment (BanzaiMan) [inline lib/travis/build/script/d.rb]: Indentation seems inconsistent here.

**Comment by BanzaiMan** [comment]
This has been deployed to org staging. Please test it.

**Comment by BanzaiMan** [comment]
Has this been tested?

**Comment by wilzbach** [comment]
> Has this been tested?

Thanks a lot for getting back so quickly and deploying this to staging.
So I activated a whole bunch of local forks to staging:

https://staging.travis-ci.org/wilzbach

image

A couple of them do test on all imaginable things, e.g

https://staging.travis-ci.org/wilzbach/Dscanner
https://staging.travis-ci.org/wilzbach/DCD
https://staging.travis-ci.org/wilzbach/dub/builds/695132

They might not all pass due to outdated local forks, but in a couple of hours this should be extensible tested with > 100 jobs.

I already looked at a couple of builds. So far it looks really good:

image

It also seems like we already recovered from our first timeout:

image

**Comment by wilzbach** [comment]
Okay I went through all error messages and they were mostly due to old projects tested with the newest compilers, [projects in another broken state](https://staging.travis-ci.org/wilzbach/libasync/jobs/695079), [problems with the DUB registry](https://staging.travis-ci.org/wilzbach/dub/jobs/695137),

I [once](https://staging.travis-ci.org/wilzbach/mir/jobs/695042) saw the following error, but I highly suspect that it has been an error in the beta as (1) that was the first release they shipped with DUB and (2) I haven't seen the error again (the current release of LDC is 1.7.0):

```
dub fetch mir-cpuid && dub build --arch "$ARCH" mir-cpuid
dyld: Library not loaded: /opt/local/lib/libcurl.4.dylib
  Referenced from: /Users/travis/dlang/ldc-1.1.0-beta2/bin/dub
  Reason: Incompatible library version: dub requires version 9.0.0 or later, but libcurl.4.dylib provides version 7.0.0
```

Since DUB 1.6.0 we the DUB binary will query mirror package registries, so the failure I observed in the build with an old DUB (bundled with DUB 1.4.0) shouldn't be any concern to you.

--- 

Here's a screenshot from vibe-core:

https://staging.travis-ci.org/wilzbach/vibe-core

image

and another from DScanner:

https://staging.travis-ci.org/wilzbach/Dscanner

image


It can't look better than that :)

File: lib/travis/build/script/d.rb (hunk 3 above)
Comment (wilzbach) [reply lib/travis/build/script/d.rb]: I saw you already merged this PR. Can I still fix this? (i.e. in a quick follow-up PR?)

**Comment by jacob-carlborg** [comment]
When can we expect this to be deployed?

**Comment by BanzaiMan** [comment]
@jacob-carlborg It's deployed already. Are you experiencing otherwise?

**Comment by wilzbach** [comment]
Wow that was awesomely quick. Huge thanks!
Looks all good here, e.g. I just started this:

https://travis-ci.org/dlang-community/D-Scanner/builds/327003178

image

image

**Comment by jacob-carlborg** [comment]
> It's deployed already. Are you experiencing otherwise?

@BanzaiMan no, just asking. Big thanks for the quick deployment 👍.

File: lib/travis/build/script/d.rb [hunk 4]
```diff
@@ -18,22 +19,34 @@ def setup
...
+          sh.echo 'DUB-related issues: https://github.com/dlang/dub/issues', ansi: :green
+
           sh.fold 'compiler-download' do
             sh.echo 'Installing compiler and dub', ansi: :yellow
 
             sh.cmd 'CURL_USER_AGENT="Travis-CI $(curl --version | head -n 1)"'
             sh.cmd <<-'EOT'.gsub(/^              /, '')
+              __mirrors=(
+                "https://dlang.org/install.sh"
+                "https://downloads.dlang.org/other/install.sh"
```
Comment (MartinNowak) [inline lib/travis/build/script/d.rb]: Nope, no https on S3, let me drop the mirror as we have no trivial way to check the install.sh.sig gpg signature.

File: lib/travis/build/script/d.rb [hunk 5]
```diff
@@ -18,22 +19,34 @@ def setup
...
+                "https://nightlies.dlang.org/install.sh"
+                "https://github.com/dlang/installer/raw/stable/script/install.sh"
```
Comment (MartinNowak) [inline lib/travis/build/script/d.rb]: I'd say this is unsafe as there is no manual verification step before deployment.
OTOH one could assume enough sec protection on Travis-CI to not worry too much about a very unlikely manipulation.

**Comment by jacob-carlborg** [comment]
I had some issues again with a job failing to download the install script. @BanzaiMan is there a possibility that Travis has some connectivity issues? This is the job [1] that failed, but I rebuild it. I though it would create a new job, but it looks like it overwrote the old job.

[1] https://travis-ci.org/jacob-carlborg/dstep/jobs/333759747

**Pull request merged**